oeps/oep-0050-hooks-extension-framework.rst

```diff
@@ -0,0 +1,235 @@
+of the changes that the authors have encountered after years of work extending
+the core in a way that is more compatible with the community at large, it is
+easier to reuse and far easier to migrate to newer versions of the codebase.
+
+This proposal draws inspiration from the cumulative experience of eduNEXT as the
+lead author, other members of the community, and from one of the largest
+extendable platforms for the web today for many years already, WordPress.
+
+.. _summary of the current options: https://github.com/edx/edx-platform/blob/master/docs/guides/extension_points.rst
+
+
+Specification
+=============
+
+The core of this proposed framework is the definition of a stable list of places
+in the life cycle of the learners registration, login, enrollment, course usage,
+and certification. Conversely, the authors life cycle will have a list of places
+during the creation and publication of courses. These places, called triggers
+will be written to the edx-platform code.
+Now, anyone that wants to extend the functionality of the platform at one of
+those places, can define a function, called an action or filter, to be called
+when the trigger is executed. The sum of a trigger and action will be referred
+to as a hook.
+
+Triggers, broadly fit into two categories:
+
+* Triggers that allow actions called at this time to affect the application flow
+  via their results. In this case, the actions called at the time will be called
+  filters.
+* Triggers that will not allow the application flow to be altered by the
+  actions called.
+
+Although the first document contains a list of the proposed hooks, we now
+proposed that the documentation for this is kept in-code at https://github.com/edx/edx-platform/tree/master/docs/guides.
+This location should contain a document that covers the design principles and
+details that did not land in the OEP as well as the final decisions made in the
+OEP. Also will index the list of hooks and their current status. It should not
+cover the history and all the previous states of the framework. It should
+reflect the latest way of thinking and examples of how to use the framework.
+
+Additionally, a directory in the same location will include individual files
+that cover each available trigger.
+
+The signature of the functions (input/output) of every trigger should be
+maintained in a stable way. The same applies for additions or removals from the
+list of public triggers, which should follow a path of deprecation first and
+preferably be given an optional hook where the same context information or code
+execution path can be accessed.
+
+The mechanism by which a trigger is connected to an action/filter will be a
+configurable python wrapper also contained in the platform core and by design
+must consider that:
+
+* Most actions will be implemented in plugins.
+* The order of execution for multiple actions/filters listening on a trigger
```

Comment:
[clarification] Where and how is the order of execution defined?

Reply:
With the separation of django signals for events this would be defined via a config dict in django settings. I left the specifics of the format out since they are more of a implementation detail.

https://github.com/edx/open-edx-proposals/pull/184/files#diff-f28146adea40eb552912482a8b24a19107252beba3e25295413bcd5ef8927628R106-R112